app/models/curriculum.rb

```diff
@@ -40,28 +40,40 @@ class Curriculum < ApplicationRecord
   # validations
   validates :name, presence: true, uniqueness: { scope: :unit_id }, length: { maximum: 255 }
   validates :programs, presence: true
-  validates :semesters_count, numericality: { greater_than_or_equal_to: 0 }
   validates :status, inclusion: { in: statuses.keys }
+  validate :check_semester
 
   # custom methods
   def build_semesters
-    return false unless valid?
+    return false if semesters_count >= number_of_semesters
 
-    divisor              = (semester_type.to_sym == :periodic ? 2 : 1)
-    number_of_semesters  = duration * divisor
-    (1..number_of_semesters.to_i).each do |sequence|
+    (1..number_of_semesters).each do |sequence|
       semesters.build(sequence: sequence,
-                      year:     (sequence.to_f / divisor).round,
+                      year:     (sequence.to_f / number_of_semesters_for_one_year).round,
                       term:     %i[spring fall][sequence % 2])
     end
   end
 
+  def number_of_semesters
+    @number_of_semesters ||= duration.to_i * number_of_semesters_for_one_year
+  end
+
   private
 
   # delegates
-  delegate :semester_type, :duration, to: :program
+  delegate :semester_type, :duration, to: :program, allow_nil: true
 
   def program
     @program ||= programs.last
   end
+
+  def number_of_semesters_for_one_year
+    (semester_type.to_s.to_sym == :periodic ? 2 : 1).to_i
```

Comment:
- Sanıyorum `semester_type` nil dönebildiğinden önce `to_s.to_sym` yapılmış, basitçe `&.to_sym` yapalım
- `to_i` de gereksiz, zaten int üretiliyor her halükarda

Sonuç olarak:

```suggestion
    semester_type&.to_sym == :periodic ? 2 : 1
```